Wrong angle in ArrayDisplay. changed phi to psi.

thomasgas opened the pull request:
I started to look at the plotting code in ArrayDisplay, since it was not working that well for point gammas and really giving random results for diffuse gammas and protons (as @vuillaut pointed out recently in #721 ).
The results of my investigation is that the angle used for plotting was the wrong one, so i changed the **phi** angle to **psi** angle: these angles are very similar for point sources (that's why the plots for point sources don't look that bad) but very different for other sources.
I'm now working on a simple algorithm for plotting the arrow in the right direction since, as pointed out by @kosack in #723 and #737  sometimes we have a 180° flip.
I analyzed the same file as @vuillaut and this is what I get using his notebook:
![protons](https://user-images.githubusercontent.com/24605557/43473080-72f50e02-94ef-11e8-8aea-2ca601db87c2.png)

thomasgas commented:
I agree @MaxNoe! 
I will change some names to more useful ones. I'll change them names using the description in ctapipe.io.containers.HillasParametersContainer  and your suggestions:
- `phi` --> `cog_polar_angle`
- `psi` --> `rot_angle_ellipse`

is that better?

MaxNoe commented:
Sounds good, but let's do another PR for this so we can discuss this separately

thomasgas commented:
so this PR can be closed and we leave the name changes to another PR.

MaxNoe commented:
I meant to disentangle the fix for the display from the change of variable names. So this PR fixes? the display and another renames variables.

thomasgas commented:
This PR just fixes the display, replacing phi with psi. we can either close it ore I can make some changes in the variables...what do you think?

MaxNoe commented:
I wouldn't close it but I also wouldn't add more changes here ;)

Let's keep it as it is and Karl can decide if he wants to merge it or if he has an idea how to fix this 180 deg offset first.

thomasgas commented:
i'm presently working on a simple algorithm to plot the arrows in the right direction, using the time gradient of the images. I'll add it to this PR asap :)

MaxNoe commented:
Ah, no I get what the problem is.

It's the classical problem of the disp approach, that you do not know, in which direction on the main shower axis the true source lies, right?

So we probably shouldn't plot arrows here but lines in both possible directions?

Also, is the direction in this view actually well-defined? The shower axis is something along the camera coordinates, how is this translated to the array top view?

MaxNoe commented:
> add arrows in the case the stereo reconstruction has been done or using the timing for each telescope
The skewness could also be used.

In generall, there could be a Classifier choosing from multiple variables, which is the correct direction, this is how it is done in the current FACT analysis.

vuillaut commented:
> The skewness could also be used.
> In generall, there could be a Classifier choosing from multiple variables, which is the correct direction, this is how it is done in the current FACT analysis.

Yes absolutely but this is still in the reconstruction part and not necesseraily related to the display.
Not that the display of the directions could be done without any reconstruction.

MaxNoe commented:
> Yes absolutely but this is still in the reconstruction part and not necesseraily related to the display.
Not that the display of the directions could be done without any reconstruction.

How that? Maybe I don't know what this plot should show.

I was under the impression, that this should show the direction to the source position.

vuillaut commented:
@MaxNoe 
Well, in my opinion, it shows the direction given by each telescope, for example, to visually check if the intersection of the lines corresponds with the shower impact position.

Maybe I should have been clearer when I said no reconstruction, I meant not "DL2 reconstruction" (therefore no stereoscopy, no reconstruction of the source, etc...). The display just needs the direction given by Hillas, done at "DL1 level" and per telescopes.

MaxNoe commented:
> Well, in my opinion, it shows the direction given by each telescope, for example, to visually check if the intersection of the lines corresponds with the shower impact position.

But is this really as simple as taking the hillas angle and plotting it on a differently oriented 2d plane with an offset? I don't think this is correct. I have to think about this.

thomasgas commented:
As you can see i added a very simple function which just plots a segment for each telescope and a point at the center (better visualization in my opinion).
With this code:
```
array = ArrayDisplay(event.inst.subarray.select_subarray("Paranal", layout))
array.set_line_hillas(hillas_try, range = 300, linewidth=3)  # linewidth is optional
```
the same event now looks something like this:
![protons_lines](https://user-images.githubusercontent.com/24605557/43574176-ca690594-9643-11e8-98bc-d7444d7888d3.png)

I just changed the names of the variables only in the functions `set_vector_hillas` and in this new one.

moralejo commented:
On 1 August 2018 at 16:08, Maximilian Nöthe <notifications@github.com>
wrote:

> Well, in my opinion, it shows the direction given by each telescope, for
> example, to visually check if the intersection of the lines corresponds
> with the shower impact position.
>
> But is this really as simple as taking the hillas angle and plotting it on
> a differently oriented 2d plane with an offset? I don't think this is
> correct. I have to think about this.
>
I think the lines should be the intersection of a reference horizontal
plane, i.e. Corsika's z=0, and the plane defined by the shower axis
estimated by the given telescope, and its mirror center.  Should be tested
also at mid-large zenith angles to make sure it is working fine.

For CTA-South the reference horizontal plane matches the plane were the
telescopes (mirrors' centers, I think) lie, but that is not the case for
the non-flat CTA-North site.  So, in the latter case, the line from a
telescope will not go exactly through the telescope's (x,y) coordinates, in
a representation like the ones in this thread.





-- 

Abelardo Moralejo Olaizola
Institut de Física d'Altes Energies
Tel : +34 931641662
Fax: +34 935811938
Avís - Aviso - Legal Notice - (LOPD) - http://legal.ifae.es

-- 
Avís -
Aviso - Legal Notice - (LOPD) - http://legal.ifae.es 
<http://legal.ifae.es/>

MaxNoe commented:
> As you can see i added a very simple function which just plots a segment for each telescope and a point at the center (better visualization in my opinion).
With this **code:**

That looks much better! Congrats. And it shows that the Hillas reconstruction works, right?

Is there a possibolity to also add the true impact point to compare?

thomasgas commented:
As @moralejo  pointed out we need to check whether this plotting works also for large zenith angles (i think the projection needs be handled better in the plotting tool, but maybe someone else can comment on that).
Here you are a plot for a diffuse gamma at La Palma (this is a really bright event...not all of them looks this good!!!)
![la_palma_diffuse_2_lsts](https://user-images.githubusercontent.com/24605557/43587613-e32152a4-966a-11e8-971f-72fec8682ee7.png)
Thanks @MaxNoe! yes that is the true impact point! The telescopes that are not pointing in "right" direction are at the edge of the camera...so it seems like the plotting is not that bad, but it needs to be checked with other simulations.

kosack commented:
Thanks all for this - it looks like a major improvement. 

For the question of arrows, we could (perhaps optionally) plot an arrow head in the direction of the spatial assymmetry.  In the future we could also include the time parameters as Thomas suggested, so if you give it both Hillas and Timing parameters, we could plot both the time gradient direction and the spatial assymmetry.

kosack commented:
The `test_array_display` unit test is still failing. Seems to be a unit problem in `ad.set_vector_hillas(hillas_dict)` when given the following as input:

```py3
   hillas_dict = {
            1: HillasParametersContainer(length=1.0 * u.m, phi=90 * u.deg),
            2: HillasParametersContainer(length=200 * u.cm, phi="95deg"),
        }
```

kosack commented:
Also, I definitely agree on the variable name change as well. Clearly they were ambiguous, since I used the wrong one!

kosack commented:
I'd suggest for the renaming PR, to be consistent in the naming like:
- `phi` --> `cog_polar_angle`
- `psi` --> `ellipse_rotation_angle`

At least to have "angle" be the last part of the name consistently

thomasgas commented:
I'll fix the PR in order to make it pass all the tests.
> For the question of arrows, we could (perhaps optionally) plot an arrow head in the direction of the spatial assymmetry. In the future we could also include the time parameters as Thomas suggested, so if you give it both Hillas and Timing parameters, we could plot both the time gradient direction and the spatial assymmetry.

@kosack  I'm now testing a script to find the time gradient and to plot it on the camera image. I'll make a new PR when it will be ready and give good results (almost ready).
